Encrypted blob store reuse DEK

albertzaharovits opened the pull request:
_EDITED 20.03.20 the repository secret is a textual password not a binary AES key_ 

_Follows the approach described in https://github.com/elastic/elasticsearch/pull/50846#issuecomment-594063727_

This builds upon the data encryption streams from #49896 to create an encrypted snapshot repository.
The repository encryption works with the following existing repository types: FS, Azure, S3, GCS (possibly works with HDFS and URL, but these are not tested).
The encrypted repository is protected by a ~256-bit AES key~ password stored on every node's keystore. The repository keys (KEK - key encryption key) are generated from the password using the PBKDF2 function, and are used to encrypt (using the [AES Wrap algorithm](https://www.ietf.org/rfc/rfc3394.txt)) other symmetric keys (referred to as DEK - data encryption keys) which are themselves used to encrypt the blobs of the regular snapshot.

The platinum license is required to snapshot to the encrypted repository, but no license is required to list or restore already encrypted snapshots.


Example of how to use the Encrypted FS Repository:
---

- pull down this branch and assemble the jar (`./gradlew :distribution:archives:no-jdk-darwin-tar:assemble`)
- similarly to the un-encrypted FS repository, specify the mount point of the shared FS in the `elasticsearch.yml` conf file (on all the cluster nodes), eg: `path.repo: ["/tmp/repo"]`
- make sure the cluster runs under a trial license (simplest configuration is to put `xpack.security.license.self_generated.type: true` in the `elasticsearch.yml` file
- ~generate the master 256-bit repository key (KEK)~
```
#openssl rand 32 > repository_keyfile.key
```
- store the ~key~ password inside the `elasticsearch.keystore`, on every cluster's node, eg for the ~`test_enc_key`~ `test_enc_pass` repository ~key~ password name:
```
#./bin/elasticsearch-keystore add-file repository.encrypted.test_enc_key.key repository_keyfile.key
./bin/elasticsearch-keystore add repository.encrypted.test_enc_pass.password
```
- start-up the cluster, and create the new encrypted repository, eg:
```
curl -X PUT "localhost:9200/_snapshot/test_enc?pretty" -H 'Content-Type: application/json' -d'
{
  "type": "encrypted",
  "settings": {
    "location": "/tmp/repo/enc",
    "delegate_type": "fs",
    "password_name": "test_enc_pass"
  }
}
'
```

Relates https://github.com/elastic/elasticsearch/pull/49896
Relates https://github.com/elastic/elasticsearch/issues/41910
Obsoletes https://github.com/elastic/elasticsearch/pull/50846 https://github.com/elastic/elasticsearch/pull/48221

albertzaharovits commented:
In addition to what has been recently discussed over at #50846 , this PR contains the following two important changes. I am confident that they are good choices, but I still want to raise them for discussion before this PR is ready for review (integration tests are a drag):

- instead of the repository password in the keystore , **the actual AES key is stored** (binary format). Using a textual password required a salt value, which must be generated by a secure random source, in order to generate the actual repository key (KEK) . It is rather complicated to broadcast the salt during repository verification, and storing the salt under its own blob adds other complications. I know this might detract from the convenience of the admin humans that prefer text, but the key can be generated, starting the the text, with `openssl enc -salt -aes-256-cbc -pass pass:secret -P` (the salt must now also be managed by the admin).
- creating the repository requires a new setting `key_name`, in addition to `delegate_type`. This is done in anticipation of the procedure to rotate the repository key. In this case, rotating the key is very simple, although quite long: 1) add a new secure key in the keystore 2) reload the keystore 3) update the repository as `read_only` 4) call a new to-be-added API which does the decryption and reencryption of the DEKs using the newly added key 5) update the repository from the old key name to the new name, and to remove the `read_only` status ; Lastly we could have a default value for `key_name` as the repository name, but changing it is still necessary for rotation.

@tvernum May you please think about these two issues, and let me know your thoughts?

albertzaharovits commented:
Here is my justification for putting the actual AES key in the `elasticsearch.keystore`, as opposed to a textual password, as has been the case until now. In a gist, it is both more complex to code, as well as inherently more insecure.

The chief reason is that it simplifies the code. It's not a great saving, but this is a sensible area of the code.
As a reference point, here is how it all works with the repository AES KEK (as a key not a password). The repository KEK is used to AES-WRAP **all** the DEKs. All wrapped DEKs are stored in a **single blob container**, each under it's own blob, where the blob container is named after the KEK Id. The KEK Id is the wrapping of a known, fixed plaintext. (This all works because AES-Wrap has no limit on the number of invocation (no nonce) and the ciphertext is deterministic. Moreover, even if it is not a design goal, AES-Wrapping of the same plaintext with different keys should produce unrelated pseudorandom permutations). The DEK Id has no semantics, it is a randomly generated UUID and prepended to each encrypted blob. The blob storing the wrapped DEK is named after the DEK Id (and again, sits under the KEK Id blob container).

To support textual passwords more code has to be added, in addition to what has been described above.
A textual password requires a randomly generated nonce, which is used to derive the key (using the PBKDF2 algorithm). In this sense, a password can be seen as a seed for a family of encryption keys. The decryption, during the restore, must indicate which key from the family to use. To do this, we can add semantics to the DEK Id, so that part of the DEK Id indicates the nonce used to generated the wrapping KEK, and the other part identifies the actual blob of the DEK (as before). The KEK id, which identifies the blob container storing all the wrapped DEKs, must be generated from a hardcoded member of the KEK family of the password. In other words, to generate the KEK Id (which identifies where to store and retrieve from wrapped DEKs), use a hardcoded nonce and a hardcoded plaintext, and use the nonce and the password to generate an AES key which is then used to AES wrap the hardcoded plaintext.
**Besides being complicated, the hardcoded nonce exposes the password to brute forcing.** And there is no convenient way around it. It is simply not enough entropy in the password to generate an identifier for it, so that it can be stored on the storage service. An identifier for the repository secret (be it password or key) is required during secret rotation. Without going into too much details about secret rotation, the repository secret Id is used when at some point the repository must be informed that it should switch from using one set of wrapped DEKs to different set, which have been wrapped with the new repository secret . The only secure way around it, is to simply **not use an identifier derived from the repository secret, and instead add another repository setting which identifies the blob container containing the set of wrapped DEKs**.
Therefore, using a password requires adding an extra repository setting, which detracts from the convenience benefits of simpler configuration.

I am willing to change my mind and adopt the textual password again, if others believe it a very important convenience feature, but for now I would continue with the repository secret key implementation if there are no strong objections against it.

CC @tvernum (the "others" in the above actually refers to you 😃 )

@nachogiljaldo do you foresee any difference one way or the other for encrypted snapshots in cloud. I believe in your case, textual or binary, repository secrets both must be generated randomly, so it shouldn't matter.

nachogiljaldo commented:
IIUC this shouldn't make a big difference on our end.

tvernum commented:
@albertzaharovits 

Let me summarise to make sure I'm understanding you correctly.

- Supporting a password in the keystore means pushing it through a KDF to produce a key, that would be the KEK
- KDFs requires a salt, which should be random
- Therefore, a password on its own is insufficient to represent a Key, you need a salt as well.
- You can't ask the user for the salt, because that's not random
- You can't store the salt in cluster state because then you can't restore into an empty cluster
- So the salt needs to be stored in the snapshot data.

At that point I'm a bit lost.
You have an "id" for a KEK, so it seems like that id could be "version n, of the secure setting named x, KDF'ed using y, salted with z" 
Then when you write out a DEK, you would include that metadata to indicate which KEK it is encrypted with.
I don't follow why that creates a big complication.

That said, _progress over perfection_. If it is substantially easier to deliver this with raw AES keys in the keystore, then let's do that.

albertzaharovits commented:
Thank you for thinking through it @tvernum !

> Let me summarise to make sure I'm understanding you correctly.
>   -  Supporting a password in the keystore means pushing it through a KDF to produce a key, that would be the KEK
>   - KDFs requires a salt, which should be random
>   - Therefore, a password on its own is insufficient to represent a Key, you need a salt as well.
>   - You can't ask the user for the salt, because that's not random
>   - You can't store the salt in cluster state because then you can't restore into an empty cluster
>   - So the salt needs to be stored in the snapshot data.

That's correct. I glanced over it, but this is indeed the core of the issue.

Next, overall I think there are two ways to look at it: either there is a single salt value, stored in the repository on the storage service, OR there are several salt values and the encrypted blobs must point to the one that was used in their case. My involved description above described the second one, because I think (but I'm not sure) that it's the simplest one to implement.

> You have an "id" for a KEK, so it seems like that id could be "version n, of the secure setting named x, KDF'ed using y, salted with z"
Then when you write out a DEK, you would include that metadata to indicate which KEK it is encrypted with.
I don't follow why that creates a big complication.

This discussion got me thinking some more, and I tried to formalize the problem.
Here are the main points:

- we can prepend anything to the encrypted blob, to help identify the associated wrapped DEK
- there can be multiple wrapped versions of the same DEK, wrapped by a different KEK
- given a KEK and a DEK Id, we should be able to find the wrapped DEK and unwrap it
- the repository KEK must be derived from the repository secret in the keystore
- the repository setting in the cluster state must identify the keystore secret used to derive the KEK (which is used to unwrap all the DEKs associated with the encrypted blobs)
- the repository settings in the cluster state must be updateable in such a way to allow switching the repository secret

That being said, I think you're right and there is enough to be able obtain the KEK from a textual password using as a nonce the random "name" of the DEK. I want to try sketch the code for it, to convince myself first, and then I'll post an update.

albertzaharovits commented:
@tvernum I've reverted back to using a password as a repository secret. It turns out we can use the DEK Id, which is generated randomly, as a nonce to compute the KEK from the repository password. This way, every DEK is wrapped by a different KEK. Generating a KEK from the password is expensive, but given the strategy to reuse DEKs, it should amortize to not be relevant in practical use cases. The changes required to move from the raw key to the password are all encapsulated in the https://github.com/elastic/elasticsearch/pull/53352/commits/406a722de0363fb57e1e733b617ec66a33ab5d04 commit (it's not a big deal, we can switch back if we have second thoughts).

From a security standpoint, weak passwords are vulnerable to brute force, because snapshots contain blobs with known plaintext, and PBKDF2 being a glorified hash, effectively turns the encrypted known plaintext into a salted hash. It's easy to see an example of it if you look into the code at how the blob name of the wrapped DEK is generated: the random DEK Id is used as the salt for the PBKDF2 function that generates the KEK, then the KEK is used to "wrap" a known plaintext. In essence, AES key are "vulnerable" to the same thing, but brute force is not something  to use when referring to random 32 bytes. Therefore textual passwords are just as secure as AES keys if they are 32 bytes long and generated randomly.

I'm still leaning for raw AES keys, if not for avoiding the PBKDF2 code shenanigans, at least for the implicit requirement that the repository secret must be unguessable (i.e. random) and long. If however we stick with passwords, maybe we should rename it to `passphrase` and specify a minimum length (currently there's none).

original-brownbear reviewed (commented):
Just some (now partly outdated) comments that we mostly already covered on another channel just now

tvernum commented:
> I'm still leaning for raw AES keys

I have no concerns with that. We may need to think about tooling at some point, but we can ship a first release that relies on external tools to generate keys.

I am starting to feel more and more as though cutting out unnecessary complexity is a valuable step to take on this project, so if raw keys does that, then it's probably the right call to make.

elasticmachine commented:
Pinging @elastic/es-security (:Security/Security)

albertzaharovits commented on plugins/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureBlobStoreRepositoryTests.java:
```diff
@@ -50,11 +50,14 @@ protected String repositoryType() {
-            .build();
+            .put(AzureStorageSettings.ACCOUNT_SETTING.getKey(), "test");
+        if (randomBoolean()) {
+            settingsBuilder.put(AzureRepository.Repository.BASE_PATH_SETTING.getKey(), randomFrom("test", "test/1"));
```
Yes, I believe we need to test the `base_path` in integ tests. The encrypted repo had to be designed to honor the base path setting of the delegated repository, while itself having a blank base path setting.

albertzaharovits commented on test/framework/src/main/java/org/elasticsearch/repositories/blobstore/ESBlobStoreRepositoryIntegTestCase.java:
```diff
@@ -81,13 +83,15 @@ protected Settings repositorySettings() {
-    protected final String createRepository(final String name, final Settings settings) {
-        final boolean verify = randomBoolean();
+    protected final String createRepository(final String name) {
+        return createRepository(name, Settings.builder().put(repositorySettings()).put(repositorySettings(name)).build(), randomBoolean());
```
I honored the original behavior, but you're right it should be `true` unless specified otherwise. I've made this change (verification `true` instead of `randomBoolean`).

albertzaharovits commented on test/framework/src/main/java/org/elasticsearch/repositories/blobstore/ESBlobStoreRepositoryIntegTestCase.java:
```diff
@@ -81,13 +83,15 @@ protected Settings repositorySettings() {
 
-    protected final String createRepository(final String name) {
-        return createRepository(name, repositorySettings());
+    protected Settings repositorySettings(String repositoryName) {
```
> add the repositoryName parameter to the existing repositorySettings() method

Done, thanks for the suggestion!

albertzaharovits commented on test/framework/src/main/java/org/elasticsearch/repositories/blobstore/ESBlobStoreRepositoryIntegTestCase.java:
```diff
@@ -97,14 +101,27 @@ protected final String createRepository(final String name, final Settings settin
+        internalCluster().getDataOrMasterNodeInstances(RepositoriesService.class).forEach(repositories -> {
+            RepositoryMissingException e = expectThrows(RepositoryMissingException.class, () -> repositories.repository(name));
+            assertThat(e.getMessage(), containsString("missing"));
+            assertThat(e.getMessage(), containsString(name));
```
> Or you could just not assert on the message at all IMO

Done, thanks again!

albertzaharovits commented on test/framework/src/main/java/org/elasticsearch/repositories/blobstore/ESBlobStoreRepositoryIntegTestCase.java:
```diff
@@ -97,14 +101,27 @@ protected final String createRepository(final String name, final Settings settin
+            assertThat(e.getMessage(), containsString(name));
+        });
+
+        return name;
```
I've removed the return value; I was trying to be consistent with the return value of the `createRepository` method.

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java:
```diff
@@ -6,11 +6,562 @@
+                    return DEKCache.computeIfAbsent(DEKId, ignored -> loadDEK(DEKId));
+                } catch (ExecutionException e) {
+                    // some exception types are to be expected
+                    if (e.getCause() instanceof IOException) {
```
I used it because the `loadDEK` method is careful to throw `IOExceptions` when storage has problems and `RepositoryException` when there are encryption problems (wrong password or tampered metadata). I've learned that `IOException`s in `readBlob` can move the repository in a corrupted state, and if we wrap the `IOException` for DEK load (which is called before reading the actual blob) that'd amount to a change of behavior in those cases.

Moreover, I don't believe the `ExecutionException` wrapping adds useful information.

albertzaharovits commented:
@original-brownbear @tvernum this is ready for review proper.

Thank you @original-brownbear for the first review round. I've acted on all your suggestions. Unfortunately code shifted around a bit since I've applied `spotless`, but otherwise there were not big changes. I've mostly restructured integ tests as we discussed.

original-brownbear commented:
@albertzaharovits @tvernum any objections to simply opening a clean PR for this (including the feature branch changes) against `master` for review and the merging it once we're through the review?
Or just merging this to the feature branch and then continuing there?

1. From a quick read over this one, I think it's very close already and there's no need for the feature branch complexity.
2. My brain hurts from trying to review a diff relative to the feature branch with the feature branch now containing the previous approach :)

albertzaharovits commented:
@original-brownbear Thanks for the prompt feedback!

The feature branch is clean, and it follows master with a few days of lag, and it does not contain previous approaches (I've not merge any of those). I prefer I keep maintaining the feature branch because it gives me the liberty of adding incremental features, most notably password rotation (but maybe others, such as password min-strength requirements, and implicit repository password name). Some integration tests might also come in a follow-up as "incremental features" (although I don't have any specific examples right now).

Yet this PR is big, and does not look "incremental", because it contains the entirety of the code for the functional `EncryptedBlobStoreRepository` plugin, without password rotation. I wasn't able to find a way to divide the main code of it into **testable** smaller pieces.

original-brownbear reviewed (commented):
Thanks @albertzaharovits I gave it a more thorough read now, looks really close + good! Thanks! Didn't find any big issues just a few small + quick points :)

albertzaharovits commented on server/src/main/java/org/elasticsearch/common/blobstore/BlobPath.java:
```diff
@@ -62,6 +63,12 @@ public BlobPath add(String path) {
         return new BlobPath(Collections.unmodifiableList(paths));
     }
 
+    public BlobPath append(BlobPath otherPath) {
```
Generally I would prefer we keep utilities like these as methods rather than functions in other places. This way we avoid duplicating code.

But I understand your dislike of creeping small utility functions, so I've removed all the utilities and instead relied on the iterator.

albertzaharovits commented on server/src/test/java/org/elasticsearch/repositories/fs/FsBlobStoreRepositoryIntegTests.java:
```diff
@@ -0,0 +1,22 @@
+ */
+package org.elasticsearch.repositories.fs;
+
+public final class FsBlobStoreRepositoryIntegTests extends FsBlobStoreRepositoryBaseIntegTestCase {
```
I have made the `repositorySettings` abstract, rather than the `repositoryType`.
It felt bizarre to have the type abstract since the type is in the class name. It's better to have the settings abstract because settings can differ a lot between classes in the same  hierarchy.

albertzaharovits replied:
Let me know if this looks OK for you.

original-brownbear replied:
> It felt bizarre to have the type abstract since the type is in the class name.

I would've renamed the class then maybe :) But let's not hold this PR up on that IMO. I think this change also makes things clear enough for now. Thanks!

albertzaharovits commented on test/framework/src/main/java/org/elasticsearch/common/settings/MockSecureSettings.java:
```diff
@@ -36,10 +37,10 @@
-    private Map<String, byte[]> files = new HashMap<>();
-    private Map<String, byte[]> sha256Digests = new HashMap<>();
-    private Set<String> settingNames = new HashSet<>();
+    private Map<String, String> secureStrings = Collections.synchronizedMap(new HashMap<>());
```
Actually they are required.
In integ tests, I change values of secure settings, which are picked up by the threads of the cluster nodes. I need to assure that a change of secure settings is visible after I restart the cluster node.

I understand your reticence of changing the infra code, so I've created `SynchronizedMockSecureSettings#SynchronizedMockSecureSettings` which synchronizes access to every method, and reverted this changes.

Let me know if you prefer back these changes.

original-brownbear replied:
I still think you don't need synchronisation here at all. In the one test that you're mutating the secure settings in, you are first setting the password, then starting the nodes? Even if on different threads, the order of operations is clear in memory and you shouldn't need synchronisation? 

I'm not so much worried about the change, I'm more worried that tests will start behaving unpredictably if we need this change. If we need it, it would imply that it's not completely clear whether or not we picked up one version or the other of the changed password setting?

albertzaharovits replied:
We've discussed this on another channel and Armin's right.

Restarting a node should obviously pick up all changes effectuated on the shared data by the test thread. I was under the impression that the node restart is like a "soft reset" where the node thread continues execution and simply reloads the configuration (No reason to assume something like that but I was convinced of it).

albertzaharovits commented on x-pack/plugin/repository-encrypted/build.gradle:
```diff
@@ -8,4 +8,23 @@ esplugin {
+    testCompile project(":test:framework")
+    // required for tests of encrypted cloud repositories
+    testCompile project(path: ':plugins:repository-gcs', configuration: 'testArtifacts')
+    testCompile(project(path: ':plugins:repository-azure', configuration: 'testArtifacts'), {
```
I don't share your concerns because we're not actually using any of those dependencies in the integ tests in the `repository-encrypted` plugin, as there are no "third party" integ tests. The `repository-encrypted` plugin needs the test classes in the `repository-azure` plugin.

If we were to use those dependencies, we could debate the fragility/messiness of it, but I'm personally relaxed about guava conflicts like these.

I'm going to mark this as "Resolved", feel free to override.

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/AESKeyUtils.java:
```diff
@@ -0,0 +1,91 @@
+    private static final byte[] KEY_ID_PLAINTEXT = "wrapping known text forms key id".getBytes(StandardCharsets.UTF_8);
+
+    public static byte[] wrap(SecretKey wrappingKey, SecretKey keyToWrap) throws NoSuchPaddingException,
+        NoSuchAlgorithmException,
```
I know, it's weird, I blame `spotless` (I have to apply the formatting for this new project).
@pugnascotia can we whine about spotless formatting to you?

original-brownbear replied:
:D ok Let's ignore this for now then. I'd +1 the whining though :)

pugnascotia replied:
It's a little trickier that you'd think to settle of a good behaviour for exception lists. But, I think I can improve this by changing the config to only wrap exception lists where necessary. It's a deviation from the general philosophy of the config, but I think it'll be an overall improvement, which is what matters.

pugnascotia replied:
I raised https://github.com/elastic/elasticsearch/pull/54710

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/AESKeyUtils.java:
```diff
@@ -0,0 +1,91 @@
+    public static SecretKey unwrap(SecretKey wrappingKey, byte[] keyToUnwrap) throws NoSuchPaddingException,
+        NoSuchAlgorithmException,
+        InvalidKeyException {
+        if (false == "AES".equals(wrappingKey.getAlgorithm())) {
```
It would probably fail, these checks serve more as documentation than anything else.
I have a slight preference to keep them in, they're not too distracting, I think.

original-brownbear replied:
>  these checks serve more as documentation than anything else.

+1 to having them as documentation, but maybe turn them into `assert`s then to make it clear that this is documentation? (just a suggestion, I'm fine with it either way)

albertzaharovits replied:
I looked into it a bit more.
Indeed **the wrapping** key must be `AES` (or `Rijndael`), otherwise the Sun security provider throws `InvalidKeyException`. **The wrapped** key algorithm does not really matter as long as the byte array of the encoded key conforms to the length requirements of the AESWrap algorithm.
Given this, I've transformed all the `IllegalArgumentExceptions` into asserts because asserts better reflect the nuance that it's more an expectation given the calling context rather than a strong argument validation.

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepositoryPlugin.java:
```diff
@@ -6,26 +6,153 @@
+public class EncryptedRepositoryPlugin extends Plugin implements RepositoryPlugin {
+    static final Logger logger = LogManager.getLogger(EncryptedRepositoryPlugin.class);
+    static final String REPOSITORY_TYPE_NAME = "encrypted";
+    static final List<String> SUPPORTED_ENCRYPTED_TYPE_NAMES = Arrays.asList("fs", "gcs", "azure", "s3");
```
No particular reason, I have simply not tested it. I understand I should look into it and have a test for it as well?

original-brownbear replied:
Let's not hold this work (and especially this PR) up on that :) I think in the final version we should maybe just support any `BlobStoreRepository`? I'd say let's just add a TODO?

albertzaharovits replied:
Sounds good adding the TODO, thanks!

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java:
```diff
@@ -6,11 +6,739 @@
+        Supplier<XPackLicenseState> licenseStateSupplier,
+        char[] repositoryPassword
+    )
+        throws GeneralSecurityException {
```
@pugnascotia This is also `spotless`'s dealing. Does it look right? I'm on the fence about it.

pugnascotia replied:
Ah, good catch - there's a separate set of parameters for formatting constructors. I'll update #54710.

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java:
```diff
@@ -6,11 +6,739 @@
+        // to assure that all participating nodes in the snapshot operation are using the same repository secret
+        snapshotUserMetadata.put(PASSWORD_ID_SALT_USER_METADATA_KEY, localRepositoryPasswordIdSalt);
+        snapshotUserMetadata.put(PASSWORD_ID_USER_METADATA_KEY, localRepositoryPasswordId);
+        logger.trace(
```
Okay, I've reverted all uses of the message supplier pattern, as none used parameters that required computation.
I tended to always use a message supplier because I can more easily ignore it if I see it part of every log statement, and it's one less decision to take when coding, but I understand others' reluctance and probably a better strategy is to default on never using the pattern.

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java:
```diff
@@ -6,11 +6,739 @@
+        // stores decrypted DEKs; DEKs are reused to encrypt/decrypt multiple independent blobs
+        this.DEKCache = CacheBuilder.<String, SecretKey>builder().setMaximumWeight(DEK_CACHE_WEIGHT).build();
+        if (isReadOnly() != delegatedRepository.isReadOnly()) {
+            throw new RepositoryException(
```
From my perspective the `IllegalStateException` type is **very informative**.

It documents assumptions when reading the code (like an assert), and if you look at a stacktrace and see one of those you know this is certainly a bug (like a 500 error not a 400, also when you look at CI test failures some exceptions are expected, while an `IllegalStateException` is clearly the source of the test failure).
In my view, the wrapper `RepositoryException` is the baddie, the callers should probably take the responsibility of wrapping with a more appropriate exception. 

I've reluctantly removed the wrapping as suggested, but I'll gladly change back if I've persuaded you.

original-brownbear replied:
Ah you're right + convinced me => TIL :) Never really thought about the status code here. Sorry for the noise, please keep it then!

albertzaharovits replied:
I have pushed 55f32d614b303e4a10f0dbc1e09da69 which reinstates the `IllegalStateException` which haven't been removed by asserts as suggested in other discussions.

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/DecryptionPacketsInputStream.java:
```diff
@@ -131,11 +144,7 @@ private int decrypt(PrefixInputStream packetInputStream) throws IOException {
         }
         // extract the nonce and the counter from the packet IV
         ByteBuffer ivBuffer = ByteBuffer.wrap(packetBuffer, 0, GCM_IV_LENGTH_IN_BYTES).order(ByteOrder.LITTLE_ENDIAN);
```
Nice suggestion, thanks!

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/SingleUseKey.java:
```diff
@@ -0,0 +1,97 @@
+        return key;
+    }
+
+    public Integer getNonce() {
```
No, good catch.

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java:
```diff
@@ -6,11 +6,739 @@
+            // reuse, but possibly generate and store a new DEK
+            final SingleUseKey singleUseNonceAndDEK = singleUseDEKSupplier.get();
+            final byte[] DEKIdBytes = singleUseNonceAndDEK.getKeyId().getBytes(StandardCharsets.UTF_8);
+            if (DEKIdBytes.length != DEK_ID_LENGTH) {
```
> This should be an assertion?

It should probably be an assertion, BUT, as an exception (pun!), I would slightly prefer this case be an `Exception` (a `RepositoryException` wrapping an `IllegalStateException`). I prefer it be an exception because of the `readBlob` and `writeBlob` symmetry; one reads a fixed length prefix while the other writes the prefix. I think it's healthy to keep this redundant check at this level, even if probably the _current code_ would fail in other places in the obverse. An assertion, in my view, signifies that the logic is broken if the assertion is not satisfied, but in this case the logic in the method works OK, but the repository gets corrupted.

> Also maybe just make getKeyId return a BytesReference so we don't have to do the string to byte[] conversion here all the time. It's kind of confusing to handle something that is a bunch of bytes as a string only to convert it to byte[] before every use?

I have pushed d49bf401c44184ec9e996ef468e86d5ec that makes key ids `BytesReference`s instead of `String`s. At first, I thought it was a good suggestion, BUT the Ids were used as strings in `read/writeBlob` methods, in constructing `BlobPath` and in logging. Looking at the code after the change it doesn't look like an improvement to me. Also because now logging does require some processing we have to introduce the message supplier pattern again... I tend to favor the original approach, but, your call.

original-brownbear replied:
Ah sorry I think there was a misunderstanding here, I was only referring to `getKeyId`, that seems to always just be a `BytesReference` and never be used as `String`? The others should stay `String` if we use them as `String`.

Also makes sense + I'm fine with keeping the exception.

albertzaharovits replied:
Okay, understood, I've pushed 87e64b3dd3bc . Thanks.

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java:
```diff
@@ -6,11 +6,739 @@
+            throw LicenseUtils.newComplianceException("encrypted snapshots");
+        }
+        Map<String, Object> snapshotUserMetadata = new HashMap<>();
+        if (userMetadata != null) {
```
Except that it can, and integ tests catch it. The empty `CreateSnapshotRequest` does not set the `userMetadata` variable. I'll keep this test.

original-brownbear replied:
Urgh, you're right :) I see now. I'll fix that, but yea we gotta keep the check for now then.

albertzaharovits commented:
Thank you for the feedback @original-brownbear . I have addressed all your suggestions, this is ready for another round.

original-brownbear reviewed (approved):
Thanks for all the iterations on this @albertzaharovits ! => LGTM, I really like the approach now :) => let's merge it to the feature branch IMO!

tvernum reviewed (commented):
I've had these comments sitting for a couple of weeks, and I really should have submited them a while ago. More coming.

tvernum commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepositoryPlugin.java:
```diff
@@ -6,25 +6,153 @@
+                    throw new IllegalArgumentException("Unsupported delegate repository type [" + DELEGATE_TYPE_SETTING.getKey() + "]");
+                }
+                if (false == getLicenseState().isEncryptedSnapshotAllowed()) {
+                    logger.warn(
+                        "Encrypted snapshots are not allowed for the currently installed license."
+                            + "Snapshots to the ["
+                            + metadata.name()
+                            + "] encrypted repository are not permitted."
+                            + "All the other operations, including restore, work without restrictions.",
+                        LicenseUtils.newComplianceException("encrypted snapshots")
```
A few small suggestions:

- I think printing the license mode will assist understanding the message.
- It needs whitespace between the joined sentences
- Our typical idiom is to use parameters rather than string concatenation.

```suggestion
                    logger.warn(new ParameterizedMessage(
                        "Encrypted snapshots are not allowed for the currently installed license [{}]."
                            + " Snapshots to the [{}] encrypted repository are not permitted."
                            + " All the other operations, including restore, work without restrictions.",
                            getLicenseState().getOperationMode().description(), metadata.name()),
                        LicenseUtils.newComplianceException("encrypted snapshots")
```

albertzaharovits replied:
Thank you.

tvernum commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/SingleUseKey.java:
```diff
@@ -0,0 +1,98 @@
+        return createSingleUseKeySupplier(keyGenerator, keyCurrentlyInUse);
+    }
+
+    // for tests use only
```
If this is for tests only, why is it referenced in the class javadoc? I'm a little confused about how this class ought to be used.

albertzaharovits replied:
This method is for test only.
What is referred to in the javadoc is the method with only one parameter.
To alleviate the confusion I have renamed the method that ought to only be used in tests from `createSingleUseKeySupplier` to `internalSingleUseKeySupplier`.

tvernum commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/SingleUseKey.java:
```diff
@@ -0,0 +1,98 @@
+
+    // for tests use only
+    static <T extends Exception> CheckedSupplier<SingleUseKey, T> createSingleUseKeySupplier(
+        CheckedSupplier<Tuple<BytesReference, SecretKey>, T> keyGenerator,
```
This feels like overuse of a `Tuple`, at the very least can we have a javadoc explaining what the elements mean?

albertzaharovits replied:
I have pushed 3ff78578721aa7dd593c2a9c3971b  which explains that the key generator generates a key and the key id.

I pondered introducing another container class for the key and the key id only (no nonce), but backed down because the existence together with the container for key, key id and nonce (`SingleUseKey`) creates confusion about when to use which.

tvernum commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/SingleUseKey.java:
```diff
@@ -0,0 +1,98 @@
+        AtomicReference<SingleUseKey> keyCurrentlyInUse
+    ) {
+        final Object lock = new Object();
+        return () -> {
```
This feels like a complicated way to produce a `Supplier` of a set of keys with a sequential nonce.
Does it really need the complexity of `AtomicReference` over a `synchronized` method? 

It feels like the general concept could be written quite simply with a class that extends `Supplier`, has an internal counter and is synchronized on `get` (or uses an `AtomicInteger` for the counter).
I presume there is a reason why you took this approach instead, but the code doesn't tell me why that is.

albertzaharovits replied:
> It feels like the general concept could be written quite simply with a class that extends Supplier, has an internal counter and is synchronized on get (or uses an AtomicInteger for the counter).

This is pretty much what the code here does; here it is without production ornaments:
```
        return () -> {
                final SingleUseKey nonceAndKey = keyCurrentlyInUse.getAndUpdate(
                    prev -> prev.nonce < MAX_NONCE ? new SingleUseKey(prev.keyId, prev.key, prev.nonce + 1) : EXPIRED_KEY
                );
                if (nonceAndKey.nonce < MAX_NONCE) {
                    return nonceAndKey;
                } else {
                    synchronized (lock) {
                        if (keyCurrentlyInUse.get().nonce == MAX_NONCE) {
                            final Tuple<BytesReference, SecretKey> newKey = keyGenerator.get();
                            keyCurrentlyInUse.set(new SingleUseKey(newKey.v1(), newKey.v2(), MIN_NONCE));
                        }
                    }
                }
        };
```

Instead of extending Supplier this uses the fact that Supplier is a `FunctionalInterface`. The internal counter is the nonce itself. Instead of synchronizing on get this uses an `AtomicReference` as an `AtomicInteger`. The remaining code is to generate a new key and reset the counter. And in this last code is where the preference for `AtomicReference` over of `AtomicInteger` shows off because it allows to atomically set the key and the nonce.

I have added a few more comments. I hope this all makes more sense now.
Do you think I should change something here?

tvernum replied:
It makes sense. I think I was missing the fact that you need to generate a new base key when the nonce rolls over. but I do think this code is unnecessarily difficult to follow.
That's a combination of trying too hard to do it all in a lambda and trying to keep the locks really small. Is that necessary?

Couldn't we have:
```
        return new CheckedSupplier<SingleUseKey, T>() {
                private Tuple<BytesReference, SecretKey> key = null;
                private int nonce = 0;
                public synchronized SingleUseKey get() throws T() {
                   if( key == null || nonce == MAX_NONCE) {
                       key = keyGenerator.get();
                       nonce = MIN_NONCE;
                   } else {
                       nonce++;
                   }
                   return new SingleUseKey(key.v1(), key.v2(), nonce);
                }
        };
```

tvernum reviewed (commented):
I made it through everything other than the crypto. I'll review that next.

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepositoryPlugin.java:
```diff
@@ -6,25 +6,153 @@
+                "Snapshotting to an encrypted repository is not permitted for the current license."
+                    + "All the other operations over the encrypted repository, eg. restore, work without restrictions.",
+                LicenseUtils.newComplianceException("encrypted snapshots")
+            );
```
Yeah, you're right, this is silly. I suppose I wanted to warn users that encrypted repos are only partially supported, before actually creating them, but that's silly because you cannot tell if encrypted repos are gonna be used at all.

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepositoryPlugin.java:
```diff
@@ -6,25 +6,153 @@
+                }
+                final Repository.Factory factory = typeLookup.apply(delegateType);
+                if (null == factory || false == SUPPORTED_ENCRYPTED_TYPE_NAMES.contains(delegateType)) {
+                    throw new IllegalArgumentException("Unsupported delegate repository type [" + DELEGATE_TYPE_SETTING.getKey() + "]");
```
Thanks!

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/DecryptionPacketsInputStream.java:
```diff
@@ -124,19 +123,19 @@ public void reset() throws IOException {
-        long packetIvCounter = ivBuffer.getLong(Integer.BYTES);
-        if (packetIvNonce != nonce) {
-            throw new IOException("Packet nonce mismatch. Expecting [" + nonce + "], but got [" + packetIvNonce + "].");
-        }
```
Removing the explicit nonce check was necessary in order to support using the same DEK for multiple encrypted blobs.

Previously, every blob had its own associated metadata blob, where we stored the nonce, which was explicitly verified here, during decryption. But now there is no place where we can associate metadata for every blob (and this is by design, because associating something to every blob incurs at least one API call and it's difficult to update).

There is no security problem if we're not explicitly checking the nonce, because the nonce is part of every packet's IV and is hence validated when the packet's authn tag is validated, implicitly by the GCM algorithm.

tvernum replied:
That makes sense.
Is there still a reason why we store it in the stream then?

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/DecryptionPacketsInputStream.java:
```diff
@@ -124,19 +123,19 @@ public void reset() throws IOException {
-            throw new IOException("Packet nonce mismatch. Expecting [" + nonce + "], but got [" + packetIvNonce + "].");
-        }
+        // extract the counter from the packet IV
+        long packetIvCounter = ByteUtils.readLongLE(packetBuffer, Integer.BYTES);
```
Fair point, it now looks like this:
```
        // extract the counter from the packet IV and validate it (that the packet is in order)
        // skips the first 4 bytes in the packet IV, which contain the encryption nonce, which cannot be explicitly validated
        // because the nonce is not passed in during decryption, but it is implicitly because it is part of the IV,
        // when GCM validates the packet authn tag
        long packetIvCounter = ByteUtils.readLongLE(packetBuffer, Integer.BYTES);
        if (packetIvCounter != counter) {
            throw new IOException("Packet counter mismatch. Expecting [" + counter + "], but got [" + packetIvCounter + "].");
        }
```

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepositoryPlugin.java:
```diff
@@ -6,25 +6,153 @@
-        // Secure settings should be readable inside this method.
+    public Map<String, Repository.Factory> getRepositories(Environment env, NamedXContentRegistry registry, ClusterService clusterService) {
+        // load all the passwords from the keystore in memory because the keystore is not readable when the repository is created
+        final Map<String, char[]> repositoryPasswordsMapBuilder = new HashMap<>();
```
The repo passwords must be left in the memory at all times, and I don't like how simple it is to blank them out when using the `SecureString`. I have nonetheless pushed 835b2db2db0c8ee9a10dbfa3fc3217fab to switch from `char[]` to `SecureString`.

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepositoryPlugin.java:
```diff
@@ -6,25 +6,153 @@
+    static final Setting.AffixSetting<SecureString> ENCRYPTION_PASSWORD_SETTING = Setting.affixKeySetting(
+        "repository.encrypted.",
+        "password",
+        key -> SecureSetting.secureString(key, null)
```
Yeah, this makes total sense! Thanks.
I have updated the issue description here https://github.com/elastic/elasticsearch/issues/41910#issue-441381154 to include a small backlog of things we need to get to before the feature branch is merged in master.

albertzaharovits commented:
@tvernum I have acted on all review points.

tvernum reviewed (commented):
I'm almost there. I still need to look at the tests for the crypto, but I don't expect anything siginificant there. 

This is really good. It all makes sense and is pretty easy to follow (given that it's the intersection of some complex crypto and complex distributed storage mgt).

Sorry it's taken me this long to get through it.

tvernum commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/AESKeyUtils.java:
```diff
@@ -0,0 +1,76 @@
+     * AES Wrap algorithm is deterministic, i.e. encryption using the same key, of the same plaintext, generates the same ciphertext.
+     * Moreover, the ciphertext reveals no information on the key, and the probability of collision of ciphertexts given different
+     * keys is statistically negligible.
+     */
```
I'm going hold off reviewing the tests until we reach a conclusion on this topic ^^
If we decide to change something here that will mean I need to do another pass anyway, and even if we decide to keep this bit as-is, I think I need to get my head around it before can review the tests well.

albertzaharovits replied:
These are good questions, and I've spent **a lot** of thinking in this area. I believe I have good answers to them.

Encrypted (wrapped) DEKs are named after the Id of the KEK that was used to wrap them (in order to use a DEK, you need to know the KEK to unwrap it; we're aiming for possibly multiple living KEKs in a given repository, so the KEK Id is a way to identify which of the KEKs (you have already obtained using PBKDF2 on the node) to use). The crucial bit here is that the KEK Id is public, as anyone that can list the repo has access to KEK Ids. In order to avoid introducing other repo settings, the only way to derive an Id for a KEK is the content of the KEK itself (i.e. we could introduce a salt as a repo setting, in the cluster state, but then when the repo is installed on a different cluster, the password, as well as the salt, must be configured).

So, the problem statement is that we have a secret AES Key (a string of bytes), generated using PBKDF2, for which we need to generate a public identifier, given no extra input.
I would say there are 3 good (engineering-grade) reasons to use AESWrap:

- crypto gurus advise against reusing the same key for different crypto operations; in this case the KEK is already used for AESWrap of DEKs.
- AESWrap is deterministic.
- there are various trusted sources that state that knowing the plaintext and the ciphertext does not reveal any information about the AES key (regardless of mode). We actually rely on this fact in other places too, because **some repo files have predictable contents**. AESWrap inherits this property, so that even if the ciphertext (KEK Id) and the plaintext (hardcoded byte string) are known, no information is exposed on the KEK (wrapping key).

tvernum replied:
Let me sit with this a bit longer.

I'm persuaded that:

1. If we use the KEK (as a key) to generate the ID then we should favour AESWrap due to wanting to restrict it to a single operation.
2. Just hashing (SHA) the KEK is a bad idea because that leaks something about the key (only really an oracle to tell whether a predicted key is the real key, but that's still something).

It still wonder whether there's a better option, but I'm not sure what it would be.

- We could HMAC the KEK, but what would we use as the key in the HMAC?
- We could SHA(AESWrap(KEK, ID_TEXT)), which reduces the plaintext risks, but I'm not sure if it's actually worth it, or whether it's just throwing in more operations "for the feels".

@jkakavas I'd appreciate any thoughts you have.

jkakavas replied:
> I'm persuaded that:
> If we use the KEK (as a key) to generate the ID then we should favour AESWrap due to wanting to restrict it to a single operation.

I'm not, I think. While it is true that there are generic warnings around this in bibliography, this stems from specific vulnerabilities in sets of algorithms from key reuse ( Same RSA private key for encryption and signing, AEC-CBC for encryption plus AES-CBC-MAC ) and is not a universal truth. Given that we control the algorithms used, I can see us coming down to a safe algorithm pair.


> Just hashing (SHA) the KEK is a bad idea because that leaks something about the key (only really an oracle to tell whether a predicted key is the real key, but that's still something).

I don't see the difference here. Given that the magic bytes that constitute the to-be-wrapped-key are stored in the source code, what stops a potential attacker from using AESWrap as an oracle for valid wrapping keys ( KEKs )? Also, assuming a malicious user has access to IDs, they have access to the repository and to DEKs, so they could also use unwrapping of DEKs as an oracle. Granted, a hash-and-check oracle is much more cost-efficient than a try-to-decrypt one. 


Alternative to the already proposed solutions, we could also use something like [HKDF](https://tools.ietf.org/html/rfc5869) with a null salt to get two keys from the KEK. One for AESWrap of DEKs and another for a HMAC to produce the id of the KEK.

albertzaharovits replied:
Thanks for thinking this through, folks!

I'm going to pile on my thoughts to maintain this as proposed:

There's indeed the risk that the contents of the encrypted repository on the cloud service can be used to brute force the keys. But, given the nature of the encrypted files, certain files or at least the headers for these files are known. We don't have any protection against this, besides making it expensive to generate the KEK (PBKDF2 cost factor). There will always be the option to brute force 256-bit AES keys.

I think we can all agree that a hash function will be more appropriate in this case. So take the properties of a cryptographic hash function:

- It's hard to find two inputs with the same output, called a collision
- It's hard to find an input matching a given output, called a pre-image

Considering the AES-Wrap encryption with the KEK key as the input (and the ciphertext of the known plaintext as the output) , the properties above are trivially met by AES-Wrap otherwise it would be vulnerable to fatal key recovery attacks.

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java:
```diff
@@ -6,11 +6,682 @@
+    private static final String RAND_ALGO = "SHA1PRNG";
+    // each snapshot metadata contains an unforgeable identifier of the repository password of the master node that started the snapshot
+    // this hash is then verified on each data node before the actual shard files snapshot, as well as on the
+    // master node that finalizes the snapshot (could be a different master node, if a master failover occurred during the snapshot)
```
You understood correctly, and indeed I've explained more of the how instead of the why.

I've changed the javadoc here to say more about the why, similar to `#validateLocalRepositorySecret` (I don't think the repetition is really detrimental).

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java:
```diff
@@ -6,11 +6,682 @@
+    // the following constants can be changed freely
+    private static final String RAND_ALGO = "SHA1PRNG";
+    // each snapshot metadata contains an unforgeable identifier of the repository password of the master node that started the snapshot
+    // this hash is then verified on each data node before the actual shard files snapshot, as well as on the
```
This choice of words is indeed murky.
I now reckon the "unforgeable identifier" terminology, instead of the more explicit "salted password hash" complicates things unnecessarily. Although not technically a "hash" (it's the ciphertext of a known plaintext, when encrypting using a key derived from the password), I've switched to using the "salted password hash" terminology consistently because it conveys the intention much clearer than the "identifier" terminology.

Let me know if this is OK with you.

tvernum replied:
As discussed in a call, we've agreed to run with the current implementation.

However this also highlighted that down the track we may decide we need to make changes to the implementation of encrypted snapshots (for this part, or some other part) and we ought to have some sort of versioning scheme so that we can (if necessary) make incompatible changes to the encryption/snapshot format but mark it in such a way that ES releases can distinguish between the 2 (or more) different formats. @albertzaharovits will think about that (as a separate piece of work).

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java:
```diff
@@ -6,11 +6,682 @@
+            final BlobPath DEKBlobPath = delegatedBasePath.add(DEK_ROOT_CONTAINER).add(DEKId);
+            logger.debug("Repository [{}] loading wrapped DEK [{}] from blob path {}", repositoryName, DEKId, DEKBlobPath);
+            final BlobContainer DEKBlobContainer = delegatedBlobStore.blobContainer(DEKBlobPath);
+            final Tuple<String, SecretKey> KEK = getKEKforDEK.apply(DEKId);
```
Sounds good to me too!

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java:
```diff
@@ -6,11 +6,682 @@
+    // as the associated encrypted DEKs)
+    private final BlobStoreRepository delegatedRepository;
+    // every data blob is encrypted with its randomly generated AES key (DEK)
+    private final Supplier<Tuple<BytesReference, SecretKey>> DEKGenerator;
```
Yeah, I think this is the code format convention we generally use.

There are many other variable names that start with caps because the first word is an acronym , and I've renamed them all.

albertzaharovits commented:
@tvernum thanks for the review, I have addressed all your points.

 f6feb6c2c845aaab31cec4b4047c77 merged in the master branch the changes required for searchable snapshots so now there's a new blob `read` function that takes an offset and a length as parameters. The `repository-encrypted-client-side` feature branch is behind this change, in order to avoid implementing that in this PR. I suggest we merge this as is, in the feature branch, and then follow-up with changes to support offset reads and decryption.
Unfortunately bwc tests fail in this situation 😔

albertzaharovits commented on x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java:
```diff
@@ -6,11 +6,694 @@
+            try {
+                // read the DEK Id (fixed length) which is prepended to the encrypted blob
+                final byte[] dekIdBytes = new byte[DEK_ID_LENGTH];
+                final int bytesRead = Streams.readFully(encryptedDataInputStream, dekIdBytes);
```
@tvernum regarding your inquiry about encrypted repo versioning:

Encrypted blobs have only the DEK Id (plaintext UUID) prepended, so the repo **currently doesn't support snapshots of multiple versions**. If we plan that future encrypted repos be backwards compatible with the current version, in a way that the new version be able to work with a repo of the current version, then we might be in trouble.  In this case, any given snapshot can technically contain blobs encrypted in multiple versions, so we have to rely on the encrypted blob itself to specify which version of the repo encryption "scheme" to use (when decrypting, i.e. snapshot restore).

I will be adding this item to the list from https://github.com/elastic/elasticsearch/issues/41910 . I concur we mustn't release before we have some way to accommodate possible future versions of the repo.

But, blob level versioning is tricky. I remember I coded serialisation of version headers tacked on encrypted blobs. Probably the version header must be authenticated, but not encrypted. All that is messy and I really appreciate the simplicity of just prepending a plaintext UUID to the encrypted blob (the current approach). 

I think prepending a "magic byte" to the UUID (which prepends the encrypted blobs) should be enough. Future versions might have a different value for the magic byte (which they might need to authenticate but not encrypt) and they might not even need one because they would infer the DEK by some other, more efficient,  means (eg. a snapshot-level dictionary of blob name to DEK content).

tvernum commented on x-pack/plugin/repository-encrypted/src/test/java/org/elasticsearch/repositories/encrypted/AESKeyUtilsTests.java:
```diff
@@ -0,0 +1,52 @@
+        SecretKey key1 = new SecretKeySpec(key1Bytes, "AES");
+        byte[] key2Bytes = randomByteArrayOfLength(AESKeyUtils.KEY_LENGTH_IN_BYTES);
+        SecretKey key2 = new SecretKeySpec(key2Bytes, "AES");
+        assertThat(AESKeyUtils.computeId(key1), not(equalTo(AESKeyUtils.computeId(key2))));
```
Should we not also verify that the id is constant? That is
```suggestion
        assertThat(AESKeyUtils.computeId(key1), not(equalTo(AESKeyUtils.computeId(key2))));
        assertThat(AESKeyUtils.computeId(key1), equalTo(AESKeyUtils.computeId(key1)));
        assertThat(AESKeyUtils.computeId(key2), equalTo(AESKeyUtils.computeId(key2)));
```

tvernum reviewed (approved):
LGTM (with 1 suggestion).
Sorry for the slow turn around

albertzaharovits committed:
Build gradle for cloud repos exports test jar

albertzaharovits committed:
Repository integ tests

albertzaharovits committed:
Barely sufficient license check

albertzaharovits committed:
Repository encrypted apply patch

albertzaharovits committed:
build.gradle modernize and "enable integ tests"

albertzaharovits committed:
Compilation problems

albertzaharovits committed:
Spotless

albertzaharovits committed:
Repository azure and repository gcs work

albertzaharovits committed:
Before moving the internal cluster tests to the dedicated source set

albertzaharovits committed:
Hooray!

albertzaharovits committed:
More than one connection open at one time for ESMockAPIBasedRepo

albertzaharovits committed:
Merge branch 'repository-encrypted-client-side-reformated' into reuse-DEKs-universally-reformated

albertzaharovits committed:
Hdfs compilation check

albertzaharovits committed:
Checkstyle

albertzaharovits committed:
Nit

albertzaharovits committed:
Update x-pack/plugin/repository-encrypted/src/test/java/org/elasticsearch/repositories/encrypted/AESKeyUtilsTests.java

Co-authored-by: Tim Vernum <tim@adjective.org>

albertzaharovits committed:
Hdfs integ tests do not preserve repo contents after delete

albertzaharovits committed:
Merge branch 'repository-encrypted-client-side-reformated' into reuse-DEKs-universally-reformated

choobinejad commented:
@albertzaharovits I see some change to the way the KEK is established. Originally it looks like this was a 256-bit AES key, and now it's:
> generated from the password using the PBKDF2 function

Can you provide more detail about where the KEK comes from? How is it generated? How is it stored? I think the origin and security of the KEK itself is a key component of this feature.

albertzaharovits commented:
@choobinejad I've created a sketch diagram in https://github.com/elastic/elasticsearch/issues/41910#issuecomment-769375228 to try to explain it. Sorry it took so long....

The KEK is generated from the password using a PBKDF2 algorithm, where the salt is the id of the DEK for which the KEK is used to wrap it; so you've got a 1-1 relationship between DEKs and KEKs. KEKs are never stored, they're only in memory as a proxy to allow using the same password to encrypt DEKs.

choobinejad commented:
Thanks @albertzaharovits, that info and the sketch are really useful!
cc @mbarretta
